### PR TITLE
Add exchange context implementation

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
Previous WML baseline port included the CHIPExchangeMgr header and cpp file, however apparently the cpp for the exchange context is separated.

 #### Summary of Changes
Added ExchangeContext.cpp from weave-core.

### CODE REVIEW

**Review by gerickson — APPROVED** [review]

File: src/lib/message/ExchangeContext.cpp [hunk 1]
```diff
@@ -0,0 +1,1664 @@
...
+    // Fill in appropriate message header flags
+    if (sendFlags & kSendFlag_DelaySend)
+        msgInfo->Flags |= kChipMessageFlag_DelaySend;
+    // FIXME: RS: possibly unnecessary, should addref instead
+    if (sendFlags & kSendFlag_RetainBuffer)
+        msgInfo->Flags |= kChipMessageFlag_RetainBuffer;
+    if (sendFlags & kSendFlag_AlreadyEncoded)
+        msgInfo->Flags |= kChipMessageFlag_MessageEncoded;
+    if (sendFlags & kSendFlag_ReuseMessageId)
+        msgInfo->Flags |= kChipMessageFlag_ReuseMessageId;
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:558]: I thought secure channel didn't want message ID reuse? @RyanTheOptimist @balducci-apple

File: src/lib/message/ExchangeContext.cpp [hunk 2]
```diff
@@ -0,0 +1,1664 @@
...
+    if (sendFlags & kSendFlag_ReuseSourceId)
+        msgInfo->Flags |= kChipMessageFlag_ReuseSourceId;
+    if (sendFlags & kSendFlag_DefaultMulticastSourceAddress)
+        msgInfo->Flags |= kChipMessageFlag_DefaultMulticastSourceAddress;
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:562]: For all of these flags, are all these features needed in CHIP? I looked in the header, and it's not clear to me what most of these do...

File: src/lib/message/ExchangeContext.cpp [hunk 3]
```diff
@@ -0,0 +1,1664 @@
...
+
+    SetFlag(msgInfo->Flags, kChipMessageFlag_FromInitiator, IsInitiator());
+
+#if CHIP_CONFIG_ENABLE_EPHEMERAL_UDP_PORT
+    SetFlag(msgInfo->Flags, kChipMessageFlag_ViaEphemeralUDPPort, UseEphemeralUDPPort());
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:567]: Similar to above, what feature is this for?

File: src/lib/message/ExchangeContext.cpp [hunk 4]
```diff
@@ -0,0 +1,1664 @@
...
+CHIP_ERROR ExchangeContext::EncodeExchHeader(ChipExchangeHeader * exchangeHeader, uint32_t profileId, uint8_t msgType,
+                                             PacketBuffer * msgBuf, uint16_t sendFlags)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    // Fill the exchange header to the message buffer.
+    exchangeHeader->Version     = kChipExchangeVersion_V1;
+    exchangeHeader->ExchangeId  = ExchangeId;
+    exchangeHeader->ProfileId   = profileId;
+    exchangeHeader->MessageType = msgType;
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:725]: How this interacts with existing CHIP security and protocolID, isn't clear to me... might be worth having folks sync up to understand how this plans to land in the existing CHIP codebase?

File: src/lib/message/ExchangeContext.cpp [hunk 5]
```diff
@@ -0,0 +1,1664 @@
...
+    // sendFlags under special circumstances (such as a retransmission
+    // of the remote alarm) can override the initiator flag in the
+    // exchange header.  The semantics here really is: use the
+    // ExchangeId in the namespace of the SourceNodeId
+    exchangeHeader->Flags = (IsInitiator() || (sendFlags & kSendFlag_FromInitiator)) ? kChipExchangeFlag_Initiator : 0;
+
+    // RMP PreProcess Checks and Flag setting
+    if (mMsgProtocolVersion == kChipMessageVersion_V2)
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:733]: I'm guessing we only need v1 here? :) We haven't shipped yet!

File: src/lib/message/ExchangeContext.cpp [hunk 6]
```diff
@@ -0,0 +1,1664 @@
...
+    return err;
+}
+
+/**
+ *  Cancel the Trickle retransmission mechanism.
+ *
+ */
+void ExchangeContext::CancelRetrans()
+{
+    // NOTE: modify for other retransmission schemes
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:778]: Should there be an issue tracking something here? (Again, sorry for the vague nature of things, not quite sure the intent of the source code)

File: src/lib/message/ExchangeContext.cpp [hunk 7]
```diff
@@ -0,0 +1,1664 @@
...
+void ExchangeContext::DoClose(bool clearRetransTable)
+{
+    // Clear app callbacks
+    OnMessageReceived       = NULL;
+    OnResponseTimeout       = NULL;
+    OnRetransmissionTimeout = NULL;
+    OnConnectionClosed      = NULL;
+    OnKeyError              = NULL;
+
+#if CHIP_CONFIG_ENABLE_RELIABLE_MESSAGING
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:804]: Should we just remove this define? We need reliable messaging in CHIP :)

File: src/lib/message/ExchangeContext.cpp [hunk 8]
```diff
@@ -0,0 +1,1664 @@
...
+
+/**
+ *  Handle trickle message within the exchange context.
+ *
+ *  @param[in]    pktInfo    A pointer to the IPPacketInfo object.
+ *
+ *  @param[in]    msgInfo    A pointer to the CHIP message info structure.
+ *
+ */
+void ExchangeContext::HandleTrickleMessage(const IPPacketInfo * pktInfo, const ChipMessageInfo * msgInfo)
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:1045]: What's a trickle message?

File: src/lib/message/ExchangeContext.cpp [hunk 9]
```diff
@@ -0,0 +1,1664 @@
...
+    uint32_t PauseTimeMillis = 0;
+#endif
+
+    // We hold a reference to the ExchangeContext here to
+    // guard against Close() calls(decrementing the reference
+    // count) by the application before the CHIP Exchange
+    // layer has completed its work on the ExchangeContext.
+    AddRef();
+
+    if (msgInfo->MessageVersion == kChipMessageVersion_V2)
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:1538]: (same as above) Probably don't need version 2 here

File: src/lib/message/ExchangeContext.cpp [hunk 10]
```diff
@@ -0,0 +1,1664 @@
...
+            if (!ShouldDropAck())
+            {
+                err = RMPHandleNeedsAck(msgInfo);
+            }
+        }
+#endif
+    }
+
+    // If the message is a duplicate and duplicates are not allowed for this exchange.
+    if ((msgInfo->Flags & kChipMessageFlag_DuplicateMessage) && !AllowDuplicateMsgs)
```
Comment (woody-apple) [inline src/lib/message/ExchangeContext.cpp:1562]: Not quite sure why CHIP would support both modes where messages and duplicated, and not... (even cryptographically). Why wouldn't we just not allow duplicates, and drop?

**Review by woody-apple — CHANGES_REQUESTED** [review]
Looks great, a few questions/comments - and things to maybe update / sync up on.

File: src/lib/message/ExchangeContext.cpp (hunk 1 above)
Comment (andy31415) [reply src/lib/message/ExchangeContext.cpp:558]: This is a straight copy from weave, no attempts at updating. I think we should update in separate PRs (lots of cleanup needed in general).

I expect the entire security model needs to be plugged out and replaced with SCTT.

Comment (yufengwangca) [reply src/lib/message/ExchangeContext.cpp:558]: @woody-apple 

This is just the strait copy of the missing parts in WML, these files are not the final version and not get linked at this moment.

We are working on integrating the existing session management logic to this baseline in separate PRs. Here is the open items list for this effort:
https://github.com/project-chip/connectedhomeip/labels/cleanup_wml_baseline

**Review by woody-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

File: src/lib/message/ExchangeContext.cpp (hunk 1 above)
Comment (robszewczyk) [reply src/lib/message/ExchangeContext.cpp:558]: Yes, this should be dropped.  This is a bit of a vestigial code that we're not pulling into CHIP (this was aimed at a self regulating group communication pattern that implemented a trickle algorithm, and we're happy to pull that out).  To that end, we should pull out: 

- The flag as identified here
- `ResendMessage` implementation (in ExchangeContext only)
- any if statements guarded with references to Trickle
- TeardownTrickleRetransmit
- StartTimerT
- StartTimerTau
- TimerT
- TimerTau
- HandleTrickleMessage
- SetupTrickleRetransmit

File: src/lib/message/ExchangeContext.cpp (hunk 8 above)
Comment (robszewczyk) [reply src/lib/message/ExchangeContext.cpp:1045]: Per my other comment, we should remove the references to the this functionality

File: src/lib/message/ExchangeContext.cpp (hunk 10 above)
Comment (robszewczyk) [reply src/lib/message/ExchangeContext.cpp:1562]: I think this was only relevant for the Trickle protocol; should be elided from the code

**Review by robszewczyk — APPROVED** [review]

**Pull request merged**